### PR TITLE
esm: scoped --type, cpp refactoring

### PR DESCRIPTION
**Pull request by guybedford**
With the current implementation, there is an edge case of `--type` that doesn't quite work out.

Consider:

main.js
```js
import './dep.js';
```

where `dep.js` is also an ECMAScript module.

Currently, `node --type=module main.js` will execute main.js as an ECMAScript module, but then when it loads dep.js it forgets the type flag and goes back to the old rules of CJS treatment, which is usually not the intention.

This PR fixes this case by having the top-level `--type` flag apply as a scope on the folder of the entry point.

I ended up refactoring the type errors and ESM_FORMAT implementation in the process, making both more consistent as a result while retaining the same implementation behaviours as previously. Moving ESM_FORMAT back into the C++ code in the process also fixes the TODO on the package.json cache sharing issue.

I've also updated the spec on how the `--type` flag works to reflect the above, and also fixed some spec bugs where it wasn't quite aligning with our implementation in that files without an extension that weren't the main were still being permitted.

### CODE REVIEW

**Comment by ljharb** [comment]
That was definitely the intention as i understood it - the flag was *only* ever supposed to apply to the entry point. Everything else follows normal disambiguation rules.

**Comment by guybedford** [comment]
Yes but we didn't discuss the dependency case, which would likely be a common catch here.

File: lib/internal/modules/esm/default_resolve.js [hunk 1]
```diff
@@ -1,22 +1,17 @@
...
 const preserveSymlinksMain = getOptionValue('--preserve-symlinks-main');
-const { ERR_INVALID_PACKAGE_CONFIG,
-        ERR_TYPE_MISMATCH,
-        ERR_UNKNOWN_FILE_EXTENSION } = require('internal/errors').codes;
-const { resolve: moduleWrapResolve } = internalBinding('module_wrap');
-const { pathToFileURL, fileURLToPath, URL } = require('internal/url');
-const asyncESM = require('internal/process/esm_loader');
-
-const realpathCache = new Map();
-// TOOD(@guybedford): Shared cache with C++
```
Comment (guybedford) [inline lib/internal/modules/esm/default_resolve.js]: This is the TODO that is fixed in the refactoring process.

**Comment by guybedford** [comment]
I've always kind of seen this one both ways, so this is very much a group decision to be considered and made I think.

The refactor was a direct consequence of the development process, which I can still split out no problem if the scoped type flag scope is rejected.

**Comment by GeoffreyBooth** [comment]
> To me, it was always clear that dependencies would be uneffected by the type flag, and I’m pretty sure we did discuss it.

Dependencies that have their _own_ scopes, like subfolders under `node_modules` that have their own `package.json` files, are unaffected by the `--type` flag. But files in the same scope of the initial entry point _need_ to be treated the same way as the initial entry point. Otherwise we have the case where, to use the example above, `main.js` is treated as ESM but `dep.js` right next to it in the same folder is treated as CommonJS. It’s never been the intent for side-by-side `.js` files to be treated as different types. That’s a bug.

**Comment by ljharb** [comment]
@GeoffreyBooth then we need to revisit `--type` at all, i guess. All files brought in by import or require should behave the same, unrelated to any flags passed to the process. The *only* reason imo for allowing `--type` in the first place is because entry points without a package.json, or extensionless files, lack the ability to otherwise provide an explicit signal.

In other words, in that example, `main.js` is ESM because of the flag, but `dep.js` is CJS. It'd need to be `dep.mjs`, or it'd need to have a package.json somehow overriding the CJS parse goal for `dep.js`, for it to be interpreted differently.

**Comment by GeoffreyBooth** [comment]
`--type` only applies to `.js` files where there’s no `package.json` `"type"` field (or the non-file cases like `--eval`). We had a long discussion about all the errors that should be thrown when `--type` conflicts with the `"type"` field or with explicit file extensions `.mjs` or `.cjs`. So `--type` only applies to `.js` files in a scope with no `"type"` field, or the non-file cases.

In the example above, if there was a defined `"type"` field _either way,_ there would be no issue. A `"type": "module"` would cause both files to be interpreted as ESM, and they’d run fine; or a `"type": "commonjs"` would cause the command `node --type=module main.js` to throw an error, as the `--type` flag would be in conflict with the `"type"` field. The `--type` flag is the third order of priority for determining the format, after the file extension and `"type"` field. If either of the other two are explicit, `--type` never applies (or causes an error to be thrown).

This was discussed most explicitly regarding `--type=auto`, which by definition can’t be used with explicit extensions or a scope with a `"type"` field (since there’d be no need for detection in those cases). The only time a user can use `auto` is a project with `.js` files and no `"type"` field, and it wouldn’t make sense for `--type=auto` to potentially tell Node to handle the initial entry point as ESM but then all the other files in the project are treated as CommonJS. The whole purpose of `--type=auto` is to set the scope based on what type it detects from the initial entry point.

This is just a bugfix. There’s never been a proposal or a PR where it was intended for `.js` files in the same scope to not be treated as the same type. The docs for `--type` don’t say “use `--type=module file.js` to run `file.js` as an ES module, but if `file.js` `import`s any other files, you also need to set `"type": "module` in the nearest parent `package.json`.” If setting `"type"` were required for any project with more than one file, then `--type` would have almost no purpose; it would only apply to non-file cases and single-file projects. That’s never been the intent.

**Comment by devsnek** [comment]
it's been my consistent understanding that `--type` was only meant to apply to the entry file. i think we need to discuss this further.

**Comment by ljharb** [comment]
Again, it’s a type, not a mode. It should not persist beyond parsing the entry point.

**Comment by ljharb** [comment]
@geoffreybooth my comment [here](https://github.com/nodejs/ecmascript-modules/pull/55#discussion_r263987810) was me asking about this very thing, and your reply said specifically it was just for the entry point.

**Comment by GeoffreyBooth** [comment]
> @GeoffreyBooth my comment [here](https://github.com/nodejs/ecmascript-modules/pull/55#discussion_r263987810) was me asking about this very thing, and your reply said specifically it was just for the entry point.

Huh? You were asking what the `isMain` variable in the `--type=auto` code meant, and I said it referred to the initial entry point, because that’s what the detection is being run on (and that only). As we’ve discussed in depth, `--type=auto` doesn’t run detection on every file in a project, because there could be many files in a project that lack `import` or `export` but were still intended to be treated as ESM. We discussed how `--type=auto` was meant to use the presence of `import` or `export` in the initial entry point to define how to treat the project. `--type=auto` would be nearly useless if it were only usable on single-file projects, and we discussed its use on multi-file projects.

**Comment by ljharb** [comment]
The intent is to use it in one-off scripts. Projects have a package.json, and don’t need a flag on the CLI.

File: lib/internal/modules/esm/default_resolve.js [hunk 2]
```diff
@@ -60,45 +37,59 @@ function resolve(specifier, parentURL) {
...
+            throw e;
+        }
+      }
+
+      // Conflict between explicit extension (.mjs, .cjs) and --type
+      if (ext === '.cjs' && asyncESM.typeFlag === 'module' ||
+          ext === '.mjs' && asyncESM.typeFlag === 'commonjs') {
+        throw new ERR_TYPE_MISMATCH(
+          entryPath, ext, asyncESM.typeFlag, 'extension');
+      }
```
Comment (jdalton) [inline lib/internal/modules/esm/default_resolve.js]: It seems valid that a `.cjs` entry module could load a `.js` file with dynamic `import()` and it be treated as ESM with the `--type=module`. These checks would prevent that wouldn't they?

Comment (guybedford) [reply lib/internal/modules/esm/default_resolve.js]: No, the `main` in this context means these checks are only done for the main entry point into Node.js, not for any top-level resolution.

Comment (jdalton) [reply lib/internal/modules/esm/default_resolve.js]: Right, the checks are done for the main which is the scenario I was describing. There is no reason to stop a `.cjs` file from being the main script when `--type=module`.

Comment (guybedford) [reply lib/internal/modules/esm/default_resolve.js]: In this model of the --type being a scope, that is something I could see, yes.

But if this doesn't land, and we stick with --type being a file indicator, the check stays.

(Note that this PR doesn't add these validations, it just refactors them)

Comment (jdalton) [reply lib/internal/modules/esm/default_resolve.js]: As part of this PR though I think those guards should be revised/removed as appropriate for the scenario.

**Comment by Fishrock123** [comment]
I think I am in favor of less "virtual package json" magic, so... -0

**Pull request closed without merging**